sled-agent/src/rack_setup/service.rs

```diff
@@ -1541,25 +1541,37 @@ pub(crate) fn build_initial_blueprint_from_sled_configs(
         let mut datasets = BTreeMap::new();
         for d in sled_config.datasets.datasets.values() {
             // Only the "Crucible" dataset needs to know the address
-            let address = sled_config.zones.iter().find_map(|z| {
```

Comment:
This change to how we determine `address` is the actual bugfix. Everything else in the PR is just refactoring machinery or test stuff.

Reply:
To check my understanding: the essence of the fix is the new `if *d.name.kind() == DatasetKind::Crucible` check, right?  Thus, we'll definitely never try to set the address of anything other than a Crucible dataset.  But further, when we later find a Crucible _zone_ on the same pool, we know it's the zone for the same dataset, not some other zone (which was another problem here before).

Reply:
Correct. Previously, this was: if there is a Crucible zone on the same zpool as this dataset (which should always be true, since we create a Crucible zone on every zpool), then set the address to the address of that Crucible zone. Now, it's: if this is a Crucible dataset, find the Crucible zone on this dataset's zpool and use its address (or `bail!` if we couldn't find it, which should never happen); if this is not a Crucible dataset, always provide `None`.